consensus/src/experimental/execution_phase.rs

```diff
@@ -18,54 +18,107 @@ use std::sync::Arc;
 /// ExecutionPhase sends the ordered blocks to the commit phase.
 ///
 
-pub type ExecutionChannelType = (
-    Vec<Block>,
-    LedgerInfoWithSignatures,
-    StateComputerCommitCallBackType,
+pub type ResetEventContentType = ();
+pub fn reset_event_content_new() -> ResetEventContentType {}
+
+pub struct ExecutionChannelType(
+    pub Vec<Block>,
+    pub LedgerInfoWithSignatures,
+    pub StateComputerCommitCallBackType,
 );
 
+impl std::fmt::Debug for ExecutionChannelType {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+impl std::fmt::Display for ExecutionChannelType {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "ExecutionChannelType({:?}, {})", self.0, self.1)
+    }
+}
+
 pub struct ExecutionPhase {
-    executor_channel_recv: Receiver<ExecutionChannelType>,
+    executor_channel_rx: Receiver<ExecutionChannelType>,
     execution_proxy: Arc<dyn StateComputer>,
-    commit_channel_send: Sender<CommitChannelType>,
+    commit_channel_tx: Sender<CommitChannelType>,
+    reset_event_channel_rx: Receiver<oneshot::Sender<ResetEventContentType>>,
+    commit_phase_reset_event_tx: Sender<oneshot::Sender<ResetEventContentType>>,
 }
 
 impl ExecutionPhase {
     pub fn new(
-        executor_channel_recv: Receiver<ExecutionChannelType>,
+        executor_channel_rx: Receiver<ExecutionChannelType>,
         execution_proxy: Arc<dyn StateComputer>,
-        commit_channel_send: Sender<CommitChannelType>,
+        commit_channel_tx: Sender<CommitChannelType>,
+        reset_event_channel_rx: Receiver<oneshot::Sender<ResetEventContentType>>,
+        commit_phase_reset_event_tx: Sender<oneshot::Sender<ResetEventContentType>>,
     ) -> Self {
         Self {
-            executor_channel_recv,
+            executor_channel_rx,
             execution_proxy,
-            commit_channel_send,
+            commit_channel_tx,
+            reset_event_channel_rx,
+            commit_phase_reset_event_tx,
         }
     }
 
+    pub async fn process_reset_event(
+        &mut self,
+        reset_event_callback: oneshot::Sender<ResetEventContentType>,
+    ) -> anyhow::Result<()> {
+        // reset the execution phase
+
+        // notify the commit phase
+        let (tx, rx) = oneshot::channel::<ResetEventContentType>();
+        self.commit_phase_reset_event_tx.send(tx).await?;
+        rx.await?;
+
+        // exhaust the executor channel
+        while self.executor_channel_rx.next().now_or_never().is_some() {}
```

Comment:
is this idiomatic way to drain a stream?

Reply:
I did some search.. But it seems a busy loop is a natural way?